sips/sip-013/sip-013-semi-fungible-token-standard.md

```diff
@@ -0,0 +1,388 @@
+to be defined as read-only.
+
+### Decimals
+
+`(get-decimals ((token-id uint)) (response uint uint))`
+
+Returns the decimal places of a token type. This is purely for display reasons,
+where external applications may read this value to provide a better user
+experience. The ability to specify decimals for a token type can be useful for
+applications that represent different kinds of assets using one SFT. For
+example, a game may have an in-game currency with two decimals and a fuel
+commodity expressed in litres with four decimals.
+
+### Token URI
+
+`(get-token-uri ((token-id uint)) (response (optional (string-ascii 256)) uint))`
+
+Returns an optional ASCII string that is a valid URI which resolves to this
+token type's metadata. These files can provide off-chain metadata about that
+particular token type, like descriptions, imagery, or any other information. The
+exact structure of the metadata is out of scope for this SIP. However, the
+metadata file should be in JSON format and should include a `sip` property
+containing a number:
+
+```JSON
+{
+	"sip": 16
+	// ... any other properties
+}
+```
+
+Applications consuming these metadata files can base display capabilities on the
+`sip` value. It should refer to a SIP number describing a JSON metadata
+standard.
+
+### Transfer
+
+`(transfer ((token-id uint) (amount uint) (sender principal) (recipient principal)) (response bool uint))`
+
+Transfer a token from the sender to the recipient. It is recommended to leverage
+Clarity primitives like `ft-transfer?` to help safeguard users. The function
+should return `(ok true)` on success or an `err` response containing an unsigned
+integer on failure. The failure codes follow the existing conventions of
+`stx-transfer?` and `ft-transfer?`.
+
+| Error code | Description                                      |
+|------------|--------------------------------------------------|
+| `u1`       | The sender has insufficient balance.             |
+| `u2`       | The sender and recipient are the same principal. |
+| `u3`       | Amount is `u0`.                                  |
+| `u4`       | The sender is not authorised to transfer tokens. |
+
+Error code `u4` is broad and may be returned under different cirumstances. For
+example, a token  contract with an allowance mechanism can return `(err u4)`
+when the `sender` parameter has no allowance for the specified token amount or
+if the sender is not equal to `tx-sender` or `contract-owner`. A token contract
+without an allowance mechanism can return `(err u4)` simply when the `sender` is
+not what is expected.
+
+Since it is possible for smart contracts to own tokens, it is recommended to
+check for both `tx-sender` and `contract-caller` as it allows smart contracts to
+transfer tokens it owns without having to resort to using `as-contract`. Such a
+guard can be constructed as follows:
+
+```clarity
+(asserts! (or (is-eq sender tx-sender) (is-eq sender contract-caller)) (err u4))
+```
```

Comment:
My understanding is that contracts _should_ use `as-contract` to issue a transfer: the goal of `as-contract` was exactly to allow contracts to behave like normal principals when trying to invoke other contracts. Is there a reason that `as-contract` should be avoided?

Reply:
Yes. Its explained in a new SIP suggestion for safe transfer of FT/NFT asset - [see this PR](https://github.com/stacksgov/sips/pull/84). 

The gist is that forcing the marketplace contract to use `as-contract` to call into the NFT contract leaves it exposed to malicious behaviour by the NFT contract (I think @LNow first pointed this out) - e.g. transferring other assets owned by the calling contract. Safe transfer requires check to be against either tx-sender or contract-caller.

Reply:
Oh, yes, I remember now. Thanks!